paasta_tools/kubernetes_tools.py

```diff
@@ -2293,12 +2297,27 @@ def get_node_affinity(self) -> Optional[V1NodeAffinity]:
             allowlist=self.get_deploy_whitelist(),
             denylist=self.get_deploy_blacklist(),
         )
+        node_selectors = self.config_dict.get("node_selectors", {})
         requirements.extend(
             raw_selectors_to_requirements(
-                raw_selectors=self.config_dict.get("node_selectors", {}),
+                raw_selectors=node_selectors,
             )
         )
 
+        # PAASTA-18198: To improve AZ balance with Karpenter, we temporarily allow specifying zone affinities per pool
+        if pool_node_affinities:
+            current_pool_node_affinities = pool_node_affinities[self.get_pool()]
+            # If the service already has a node selector for a zone, we don't want to override it
+            if (
+                current_pool_node_affinities
+                and "topology.kubernetes.io/zone" not in node_selectors
```

Comment:
I'm wondering if `"topology.kubernetes.io/zone"` is the only label we should care about? 
Following labels include info about AZ:

- label_failure_domain_beta_kubernetes_io_zone
- label_karpenter_sh_nodepool
- label_topology_ebs_csi_aws_com_zone
- label_vpc_amazonaws_com_eni_config
- label_yelp_com_eni_config
- label_yelp_com_habitat

If the service owner adds one of them as node_selector our pod manifests will be inconsistent. Question is should we add them to the filter list?